File: 100_program_list/p34.c
// Print a Leaf Nodes in Binary search tree
#include <stdio.h>
#include <stdlib.h>

// Define a structure for the Binary Search Tree node
struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

// Function to create a new node
struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Function to insert a node in the BST
struct Node* insert(struct Node* root, int data) {
    if (root == NULL) {
        return createNode(data);
    }
    
    if (data < root->data) {
        root->left = insert(root->left, data);
    } else {
        root->right = insert(root->right, data);
    }

    return root;
}

// Function to print the leaf nodes of the BST
void printLeafNodes(struct Node* root) {
    if (root == NULL) {
        return;
    }

    // If the node is a leaf (both left and right children are NULL)
    if (root->left == NULL && root->right == NULL) {
        printf("%d ", root->data);  // Print the leaf node
    }

    // Recur for left and right subtrees
    printLeafNodes(root->left);
    printLeafNodes(root->right);
}

int main() {
    struct Node* root = NULL;

    // Insert nodes into the BST
    root = insert(root, 50);
    root = insert(root, 30);
    root = insert(root, 20);
    root = insert(root, 40);
    root = insert(root, 70);
    root = insert(root, 60);
    root = insert(root, 80);

    // Print the leaf nodes
    printf("Leaf Nodes in the Binary Search Tree: ");
    printLeafNodes(root);
    printf("\n");

    return 0;
}

File: 100_program_list/p34.java
// Java program to print the leaf nodes of a Binary Search Tree

class p34 {
    
    // Define the structure of the Node
    class Node {
        int data;
        Node left, right;

        public Node(int data) {
            this.data = data;
            left = right = null;
        }
    }
    
    // Root of the Binary Search Tree
    Node root;

    // Constructor
    public p34() {
        root = null;
    }

    // Function to insert a new node in the BST
    public Node insert(Node root, int data) {
        // If the tree is empty, create a new node
        if (root == null) {
            root = new Node(data);
            return root;
        }

        // Otherwise, recur down the tree
        if (data < root.data)
            root.left = insert(root.left, data);
        else if (data > root.data)
            root.right = insert(root.right, data);

        // return the (unchanged) node pointer
        return root;
    }

    // Function to print all leaf nodes in the Binary Search Tree
    public void printLeafNodes(Node root) {
        if (root == null) {
            return;
        }

        // If both left and right children are null, it's a leaf node
        if (root.left == null && root.right == null) {
            System.out.print(root.data + " ");
        }

        // Recur for left and right subtrees
        printLeafNodes(root.left);
        printLeafNodes(root.right);
    }

    // Main method to test the functionality
    public static void main(String[] args) {
        p34 tree = new p34();

        // Insert nodes into the Binary Search Tree
        tree.root = tree.insert(tree.root, 50);
        tree.root = tree.insert(tree.root, 30);
        tree.root = tree.insert(tree.root, 20);
        tree.root = tree.insert(tree.root, 40);
        tree.root = tree.insert(tree.root, 70);
        tree.root = tree.insert(tree.root, 60);
        tree.root = tree.insert(tree.root, 80);

        // Print leaf nodes in the tree
        System.out.println("Leaf Nodes in the Binary Search Tree:");
        tree.printLeafNodes(tree.root);
    }
}
